File: files/CSEC 23032025.R
library(tidyverse)
library(readxl)

path = "files/Challenge1225.xlsx"
input = read_excel(path, range = "B3:C7")
test  = read_excel(path, range = "E3:F7")

result = input %>%
  separate_rows(Narration, sep = "[,.]") %>%
  mutate(action = case_when(
    str_detect(Narration, "sell|sold") ~ "sell",
    str_detect(Narration, "buy") ~ "buy",
    TRUE ~ NA_character_
  ))  %>%
  na.omit() %>%
  mutate(amount = as.numeric(str_extract(Narration, "\\d+")) * ifelse(action == "buy", -1, 1)) %>%
  summarise(`Profit (loss)` = sum(amount), .by = Items)

all.equal(result, test, check.attributes = FALSE)
#> [1] TRUE

File: files/2025-03-23/CSEC 23032025.py
import pandas as pd
import re

path = "files/Challenge1225.xlsx"
input = pd.read_excel(path, usecols="B:C", skiprows=2, nrows=5)
test = pd.read_excel(path, usecols="E:F", skiprows=2, nrows=5).rename(columns=lambda col: re.sub(r'\.1$', '', col))

input = input.assign(
    Narration=input['Narration'].str.split(r'[,.]')
).explode('Narration').dropna(subset=['Narration'])

input['action'] = input['Narration'].str.extract(r'(sell|sold|buy)', expand=False)
input = input.dropna(subset=['action'])
input['amount'] = input['Narration'].str.extract(r'(\d+)').astype('int64')
input.loc[input['action'] == 'buy', 'amount'] *= -1

result = input.groupby('Items', sort=False)['amount'].sum().rename('Profit (loss)').reset_index()

print(result.equals(test)) # True
